lib/active_record/virtual_attributes/virtual_fields.rb

```diff
@@ -136,102 +143,70 @@ class Base
   module Associations
     class Preloader
       prepend(Module.new {
-        # preloader.rb active record 6.0
-        # changed:
-        # since grouped_records can return a hash/array, we need to handle those 2 new cases
-        def preloaders_for_reflection(reflection, records, scope, polymorphic_parent)
-          case reflection
-          when Array
-            reflection.flat_map { |ref| preloaders_on(ref, records, scope, polymorphic_parent) }
-          when Hash
-            preloaders_on(reflection, records, scope, polymorphic_parent)
-          else
-            super(reflection, records, scope)
-          end
-        end
-
-        # rubocop:disable Style/BlockDelimiters, Lint/AmbiguousBlockAssociation, Style/MethodCallWithArgsParentheses
-        # preloader.rb active record 6.0
-        # changed:
-        # passing polymorphic around (and makes 5.2 more similar to 6.0)
-        def preloaders_for_hash(association, records, scope, polymorphic_parent)
-          association.flat_map { |parent, child|
-            grouped_records(parent, records, polymorphic_parent).flat_map do |reflection, reflection_records|
-              loaders = preloaders_for_reflection(reflection, reflection_records, scope, polymorphic_parent)
-              recs = loaders.flat_map(&:preloaded_records).uniq
-              child_polymorphic_parent = reflection && reflection.respond_to?(:options) && reflection.options[:polymorphic]
-              loaders.concat Array.wrap(child).flat_map { |assoc|
-                preloaders_on assoc, recs, scope, child_polymorphic_parent
-              }
-              loaders
+        # preloader is called with virtual attributes - need to resolve
+        def call
```

Comment:
where is `call` being called?

Reply:
Great question.

Sure, we do explicitly call it in our specs: https://github.com/kbrock/activerecord-virtual_attributes/blob/allow_rails_7/spec/virtual_includes_spec.rb#L542

But most of the time, it is called directly by rails.

From what I can tell, this is mostly called directly from preload_associations via exec_query.

https://github.com/rails/rails/blob/7-0-stable/activerecord/lib/active_record/relation.rb#L825-L832

I did try patching in that caller, but when we ran the preloader directly, this had issues - as you know since you had to skip those 3 tests before (we were since able to unskip).

So this fixes 18 specs (probably 50+ uses) in virtual_includes_specs like:

```
expect(Author.includes(:nick_or_name)).to preload_values(:nick_or_name, author_name)
```

This `includes` value ends up in a Preloader. As you know, it is not valid. So we need to translate from the old includes values to the proper ones.

Now I did try and just update `@associations` value directly, but that only worked if there were no additional levels, so many tests were still failing.

This fix called allows us to get more complicated with our associations.
Coincidently, this recursive-ish solution is similar to the v6 changes where we call into self if the value changed. That was purely coincidental but brought me comfort.